### DISCUSSION TITLE
Is it possible to launch an app from the CLI?

### DISCUSSION THREAD

**Discussion by mattleibow** [microsoft/WindowsAppSDK#2391] [Q&A]
I have been spoilt with `dotnet run` all these years, so I was wondering is there a set of CLI args that can be used to launch a Windows App from the CLI?

This can be with either msbuild or dotnet. I know of a few ways that don't seem to work:

1. `msbuild /t:Run`
2. `dotnet run`
3. `dotnet build -t:Run`

Not sure if maybe there is a special task for that?

Also, how does this work with packaged vs unpackaged apps?

From @eerhardt

I can get an unpackaged app to launch from the command line:

```xml
<WindowsPackageType>None</WindowsPackageType>
<WinUISDKReferences>false</WinUISDKReferences>
```

but you need to build first, and then run

```   
$ dotnet build .\Microsoft.NetConf2021.Maui.csproj -f net6.0-windows10.0.19041.0 -c Release
$ dotnet build .\Microsoft.NetConf2021.Maui.csproj -t:Run -f net6.0-windows10.0.19041.0 -c Release
```

**Comment by DrusTheAxe** [comment]
Is your app packaged or unpackaged?

If packaged then you need to know your application's Application User Model ID (AUMID). That's a derived value of `PackageFamilyName + "!" + PackageRelativeApplicationID`. The latter (PRAID) is what you specified in your `appxmanifest.xml` as `<Application Id="...praid..."...>`. Once you know your aumid you can launch it via `shell:AppsFolder` e.g. `shell:AppsFolder\Microsoft.WindowsCalculator_8wekyb3d8bbwe!App`

There's oodles of APIs and recipes to discover AUMIDs for installed apps as well as compute it arbitrary values (w/o anything installed e.g. [FormatApplicationUserModelId()](https://docs.microsoft.com/windows/win32/api/appmodel/nf-appmodel-formatapplicationusermodelid).

If unpackaged then nothing's changed. You still launch however you launch without WinAppSDK or MSIX e.g. if your app's an exe then run the exe.

**Comment by jonwis** [comment]
For packaged apps, there's also the `windows.appExecutionAlias` extension - see https://github.com/AdamBraden/AppExecutionAlias or https://stackoverflow.com/questions/55426605/how-to-properly-set-appexecutionalias-so-the-program-could-be-launched-from-comm for samples.

**Reply by jonwis** [reply]
... but note that doing so - or using @DrusTheAxe 's solution, your app needs to be deployed/installed first.

**Reply by DrusTheAxe** [reply]
P.S. For packaged apps...

...as of Win11 you can also run the exe (CreateProcess, ShellExecute, command-line, Win-R, whatever). Windows now can determine `"C:\Program Files\WindowsApps\LolzCatzVideoz_x86_1.2.3.4__1234567890abc\kittykitty.exe"` corresponds to a packaged `<Application Executeable="kittykitty.exe"...>` and launch the app. But that path changes over time (e.g. when package's version changes) so it's not stable over time or as convenient for end-user interactive use. AppExecutionAlias'es are a stable and convenient equivalent.

Alternatively, as of 20H1, you can specify `<msix...>` in an application manifest embedded in your exe and running it does the same -- Windows launches the app. Search the net for "sparse package msix" for more details, e.g. [Add the package identity metadata to your desktop application manifest](https://docs.microsoft.com/windows/apps/desktop/modernize/grant-identity-to-nonpackaged-apps#add-the-package-identity-metadata-to-your-desktop-application-manifest). But again, requires your target the exe which is typically in a path that's not stable over time i.e. AppExecutionAlias is usually more convenient.

Regardless, if you're a packaged app you need to be installed for these to work.

Unless you choose to go the 'sparse package' route then you'd be just some exe that, when run, checks if it has package identity and if not, register your package then restart and poof! Your process has your package identity. You can think of this as a hybrid approach, or a "just-in-time registration" model. Search the net for 'sparse package msix' for hits like [Grant identity to non-packaged desktop apps](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/grant-identity-to-nonpackaged-apps) for more details.
